src/adaptors/adaptor_tls.c

```diff
@@ -0,0 +1,630 @@
+                //
+                qd_log(tls->log_source,
+                       QD_LOG_TRACE,
+                       "[C%" PRIu64 "] qd_tls_decrypt raw buffer size=0, did not sent this buffer to TLS, freeing it",
+                       tls->conn_id);
+                qd_adaptor_buffer_free(adaptor_buff);
+            }
+        }
+
+        //
+        // Process TLS.
+        //
+        if (!process_tls(tls)) {
+            take_back_input_decrypt_buffs(tls);
+            return QD_TLS_ERROR;
+        }
+
+        qd_log(tls->log_source, QD_LOG_TRACE, "[C%" PRIu64 "] qd_tls_decrypt process_tls successful", tls->conn_id);
+
+    give_decrypt_output_buffers:
+        while (pn_tls_need_decrypt_output_buffers(tls->tls_session)) {
+            //
+            // Give one raw buffer at a time to tls which it will be used to decrypt the encrypted data.
+            // This loop will give enough buffers needed by tls to decrypt the data.
+            //
+            pn_raw_buffer_t      decrypted_raw_buffer;
+            qd_adaptor_buffer_t *decrypted_adaptor_buf = qd_adaptor_buffer_raw(&decrypted_raw_buffer);
+            (void) decrypted_adaptor_buf;
+
+            size_t given_decrypt_output_buffers =
+                pn_tls_give_decrypt_output_buffers(tls->tls_session, &decrypted_raw_buffer, 1);
+            (void) given_decrypt_output_buffers;  // prevent unused variable warning
+
+            //
+            // Process TLS and log an error if any.
+            //
+            if (!process_tls(tls)) {
+                take_back_input_decrypt_buffs(tls);
+                return QD_TLS_ERROR;
+            }
+
+            qd_log(tls->log_source,
+                   QD_LOG_TRACE,
+                   "[C%" PRIu64
+                   "] qd_tls_decrypt (pn_tls_need_decrypt_output_buffers) process_tls successful, "
+                   "given_decrypt_output_buffers=%zu",
+                   tls->conn_id,
+                   given_decrypt_output_buffers);
+        }
+
+        pn_raw_buffer_t decrypted_output_buff;
+        while (pn_tls_take_decrypt_output_buffers(tls->tls_session, &decrypted_output_buff, 1)) {
+            qd_adaptor_buffer_t *decrypted_adaptor_buff =
+                qd_get_adaptor_buffer_from_pn_raw_buffer(&decrypted_output_buff);
+            DEQ_INSERT_TAIL(*decrypted_buffs, decrypted_adaptor_buff);
+            qd_log(tls->log_source,
+                   QD_LOG_TRACE,
+                   "[C%" PRIu64 "] qd_tls_decrypt pn_tls_take_decrypt_output_buffers, decrypt adaptor buffer size=%zu",
+                   tls->conn_id,
+                   qd_adaptor_buffer_size(decrypted_adaptor_buff));
+        }
+
+        // You can take only a maximum of decrypt_output_buff_capacity. You will have to recheck
+        // pn_tls_need_decrypt_output_buffers to get everything back.
+        if (pn_tls_need_decrypt_output_buffers(tls->tls_session)) {
+            goto give_decrypt_output_buffers;
```

Comment:
Oh no - please don't use goto for a looping construct.

The only permissible use of goto's in this codebase is to jump to an exit point of a function where resources may be cleaned up on error.

Reply:
After our discussion, you mentioned it is ok to keep the goto statements, so I have not removed them.